Pubkeyinfo

HansH111 opened the pull request:
This allows you to get the pubkey info in environment var SSH_PUBKEYINFO when a ssh key is used to gain access.
So for instance:
ssh_rsa AAAAB3Nz....DI0XQ== hans1_key
ssh_rsa AAAAB3Nz....XF7DF== appl1_key
After getting access with the second key, you get in SSH_PUBKEYINFO="appl1_key"
Very handy to recognize the remote appl/user when performing automatic actions.

HansH111 commented:
I will find the problem tomorrow... currently no time...

mkj commented:
This looks useful, thanks.

I suspect the leak occurs because checkpubkey() gets called multiple times - first when the client probes usable keys, then for the actual signature. Just `m_free()` before assigning the option should be sufficient, it's `NULL` to start. 

OpenSSH has a similar [`ExposeAuthInfo`](https://man.openbsd.org/sshd_config#ExposeAuthInfo) option off by default, though that points `SSH_AUTH_INFO` to a file with details. Your approach here seems fine since Dropbear only supports one auth method at present. I'm a bit surprised there isn't a more widely used env var for this.

mkj commented:
Thanks!

The BuildTest seemed to intermittently fail due to bound sockets, I pushed a hopeful fix yesterday

mkj commented:
I had missed that the actual failure was `buf_getptr()` failing, the buffer was  truncated prior to use.
I've reworked it a bit now, and have also added an allow list of shell characters - I've seen situations where shell scripts will pass all environment variables around without escaping, that could be hazardous. The current set will allow the standard comments that ssh-keygen generates.

https://github.com/mkj/dropbear/commit/10875e8524b9d38c0ac278c872de697e416ebfac

HansH111 commented:
I just saw that to...  nice update... thx


On Wed, Mar 16, 2022 at 10:30 AM Matt Johnston ***@***.***>
wrote:

> I had missed that the actual failure was buf_getptr() failing, the buffer
> was truncated prior to use.
> I've reworked it a bit now, and have also added an allow list of shell
> characters - I've seen situations where shell scripts will pass all
> environment variables around without escaping, that could be hazardous. The
> current set will allow the standard comments that ssh-keygen generates.
>
> 10875e8
> <https://github.com/mkj/dropbear/commit/10875e8524b9d38c0ac278c872de697e416ebfac>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/mkj/dropbear/pull/151#issuecomment-1068913161>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAJPBIKNXHZVLEFHQMSAJ73VAGSZNANCNFSM5QWLOGLA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

HansH111 commented:
How about also allowing a space character, but limit the length to say max 40 characters...
This will prevent very long comments

mkj commented:
I thought about that, but escaping in shell scripting always seems a
hassle, a "single word" identifier seems easier to use? 

Also I think I've sorted out the cifuzz leak failure, waiting for it to
run through now. 

On 2022-03-16 6:44 pm, HansH111 wrote:

> How about also allowing a space character, but limit the length to say max 40 characters...
> This will prevent very long comments 
> 
> --
> Reply to this email directly, view it on GitHub [1], or unsubscribe [2].
> Triage notifications on the go with GitHub Mobile for iOS [3] or Android [4]. 
> You are receiving this because you modified the open/close state.Message ID: ***@***.***>

 

Links:
------
[1] https://github.com/mkj/dropbear/pull/151#issuecomment-1068982587
[2]
https://github.com/notifications/unsubscribe-auth/AACQDSN7HMXHP32OSHU7DXLVAG3QFANCNFSM5QWLOGLA
[3]
https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&amp;mt=8&amp;pt=524675
[4]
https://play.google.com/store/apps/details?id=com.github.android&amp;referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub

HansH111 commented:
yes, you are right, a single word identifier should be good enough.
